geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/attribute/AttributeIndexKeySpace.scala

```diff
@@ -0,0 +1,281 @@
+      AttributeIndexKey.encodeForIndex(feature.getAttribute(i), list).map(v => AttributeIndexKey(i.toShort, v))
+    }
+  }
+
+  override def toIndexKeyBytes(sft: SimpleFeatureType, lenient: Boolean): ToIndexKeyBytes = {
+    val indexedAttributes = SimpleFeatureTypes.getSecondaryIndexedAttributes(sft).map { d =>
+      val i = sft.indexOf(d.getName)
+      (i, AttributeIndexKey.indexToBytes(i), d.isList)
+    }
+    (prefix, feature, suffix) => {
+      val baseLength = prefix.map(_.length).sum + suffix.length + 3 // 2 for attributed i, 1 for null byte
+      indexedAttributes.flatMap { case (idx, idxBytes, list) =>
+        AttributeIndexKey.encodeForIndex(feature.getAttribute(idx), list).map { encoded =>
+          val value = encoded.getBytes(StandardCharsets.UTF_8)
+          // create the byte array - allocate a single array up front to contain everything
+          val bytes = Array.ofDim[Byte](baseLength + value.length)
+          var i = 0
+          prefix.foreach { p => System.arraycopy(p, 0, bytes, i, p.length); i += p.length }
+          bytes(i) = idxBytes(0)
+          i += 1
+          bytes(i) = idxBytes(1)
+          i += 1
+          System.arraycopy(value, 0, bytes, i, value.length)
+          i += value.length
+          bytes(i) = ByteArrays.ZeroByte
+          System.arraycopy(suffix, 0, bytes, i + 1, suffix.length)
+          bytes
+        }
+      }
+    }
+  }
+
+  override def getIndexValues(sft: SimpleFeatureType, filter: Filter, explain: Explainer): AttributeIndexValues[Any] = {
+    val attribute = {
+      val names = FilterHelper.propertyNames(filter, sft)
+      require(names.lengthCompare(1) == 0,
+        s"Couldn't extract single attribute name from filter '${filterToString(filter)}'")
+      names.head
+    }
+
+    val i = sft.indexOf(attribute)
+
+    require(i != -1, s"Attribute '$attribute' from filter '${filterToString(filter)}' does not exist in '$sft'")
+
+    val descriptor = sft.getDescriptor(i)
+    val binding = if (descriptor.isList) { descriptor.getListType() } else { descriptor.getType.getBinding }
+    val bounds = FilterHelper.extractAttributeBounds(filter, attribute, binding)
+
+    if (bounds.isEmpty) {
+      // we have an attribute, but weren't able to extract any bounds
+      logger.warn(s"Unable to extract any attribute bounds from: ${filterToString(filter)}")
+    }
+
+    AttributeIndexValues(attribute, i, bounds.asInstanceOf[FilterValues[Bounds[Any]]], binding.asInstanceOf[Class[Any]])
+  }
+
+  override def getRanges(values: AttributeIndexValues[Any]): Iterator[ScanRange[AttributeIndexKey]] = {
+
+    import AttributeIndexKey.encodeForQuery
+    import org.locationtech.geomesa.filter.WILDCARD_SUFFIX
+
+    val AttributeIndexValues(_, i, fb, binding) = values
+
+    if (fb.isEmpty) {
+      // we have an attribute, but weren't able to extract any bounds... scan all values
+      Iterator.single(UnboundedRange(AttributeIndexKey(i.toShort, null, inclusive = false)))
```

Comment:
When would we want to do a full-table scan on the attribute index?

Reply:
we probably wouldn't... but we could if we wanted to! :)